spec/controllers/prompts_controller_spec.rb

```diff
@@ -131,6 +131,74 @@
       it_redirects_to_simple("#{collection_signups_path(open_signup.collection)}/"\
                       "#{open_signup.collection.prompts.first.challenge_signup_id}/edit")
     end
+
+    context "prompt has tags" do
+      before do
+        # Delete existing prompt of current user
+        open_signup.offers.first.destroy
+      end
+      let!(:canonical_character) { create(:canonical_character, name: "Sakura Kinomoto") }
+
+      it "should accept canonical tags" do
+        post :create,
+          params: {
+            collection_id: open_signup.collection.name,
+            prompt_type: "offer",
+            prompt: {
+              description: "This is a description.",
+              tag_set_attributes: {
+                character_tagnames: ["Sakura Kinomoto"]
+              }
+            }
+          }
+        it_redirects_to_with_notice(
+          collection_signup_path(open_signup.collection, open_signup),
+          "Prompt was successfully added."
+        )
+      end
+
+      it "should error if some tags aren't canonical" do
+        post :create,
+          params: {
+            collection_id: open_signup.collection.name,
+            prompt_type: "offer",
+            prompt: {
+              description: "This is a description.",
+              tag_set_attributes: {
+                character_tagnames: ["Sakura Typomoto"]
+              }
+            }
+          }
+        it_redirects_to_with_error(
+          edit_collection_signup_path(open_signup.collection, open_signup),
+          "That prompt would make your overall sign-up invalid, sorry."
+        )
+        expect(assigns[:prompt].errors[:base]).to eq(
+          ["^These character tags in your offer are not canonical and cannot be used in this challenge: Sakura Typomoto. To fix this, please ask your challenge moderator to set up a tag set for the challenge. New tags can be added to the tag set manually by the moderator or through open nominations."]
+        )
+      end
+
+      it "shouldn't make needless checks on unused kind of tags" do
+        expect_any_instance_of(PromptRestriction).to receive(:has_tags?).with("character").at_least(:once)
```

Comment:
Not sure where the second call to this method happens actually, might be worth investigating.

Reply:
Well, simply because the validations are called a first time through [valid?](https://github.com/otwcode/otwarchive/blob/4e56a347944a6fb831414d9cccdfc866ddfacce5/app/controllers/prompts_controller.rb#L133) then a second time on [save](https://github.com/otwcode/otwarchive/blob/4e56a347944a6fb831414d9cccdfc866ddfacce5/app/controllers/prompts_controller.rb#L136).

Let's try to see if we can run them all only once without breaking everything.

Reply:
I was also looking at this. My in-progress comment:

> The validation is called twice in `prompts_controller#create`, once by [@challenge_signup.valid?](https://github.com/otwcode/otwarchive/blob/4e56a347944a6fb831414d9cccdfc866ddfacce5/app/controllers/prompts_controller.rb#L133) and once by [@prompt.save](https://github.com/otwcode/otwarchive/blob/4e56a347944a6fb831414d9cccdfc866ddfacce5/app/controllers/prompts_controller.rb#L136).

> This happens because challenge_signup uses `accepts_nested_attributes_for :offers, :prompts, :requests`.  [accepts_nested_attributes_for](https://api.rubyonrails.org/v6.1.7.4/classes/ActiveRecord/NestedAttributes/ClassMethods.html) means the associated record is autosaved and thus validated if the parent is validated: https://api.rubyonrails.org/v6.1.7.4/classes/ActiveRecord/AutosaveAssociation.html.

> So, [prompts_controller#create](https://github.com/otwcode/otwarchive/blob/4e56a347944a6fb831414d9cccdfc866ddfacce5/app/controllers/prompts_controller.rb#L136) can use `@prompt.save(validate: false)` instead.

Your solution of combining it also looks okay, but I'm a bit hesitant to call save on the signup when it's the prompt that is getting created.